src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+                captureToken = CreateMissingToken(RegexKind.CaptureNameToken);
+
+                if (_currentToken.Kind == RegexKind.MinusToken)
+                {
+                    return ParseBalancingGrouping(
+                        openParenToken, questionToken, openToken, captureToken.Value);
+                }
+                else
+                {
+                    openParenToken = openParenToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                        WorkspacesResources.Invalid_group_name_Group_names_must_begin_with_a_word_character,
+                        GetTokenSpanIncludingEOF(_currentToken)));
+
+                    // If we weren't at the end of the text, go back to before whatever character
+                    // we just consumed.
+                    MoveBackBeforePreviousScan();
+                }
+            }
+
+            var capture = captureToken.Value;
+            if (capture.Kind == RegexKind.NumberToken && (int)capture.Value == 0)
+            {
+                capture = capture.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Capture_number_cannot_be_zero,
+                    capture.GetSpan()));
+            }
+
+            // Can't have any trivia between the elements in this grouping header.
+            ConsumeCurrentToken(allowTrivia: false);
+
+            if (_currentToken.Kind == RegexKind.MinusToken)
+            {
+                // Have  (?<...-  parse out the balancing group form.
+                return ParseBalancingGrouping(
+                    openParenToken, questionToken,
+                    openToken, capture);
+            }
+
+            var closeToken = ParseCaptureGroupingCloseToken(ref openParenToken, openToken);
+
+            return new RegexCaptureGroupingNode(
+                openParenToken, questionToken,
+                openToken, capture, closeToken,
+                ParseGroupingEmbeddedExpression(_options), ParseGroupingCloseParen());
+        }
+
+        private RegexToken ParseCaptureGroupingCloseToken(ref RegexToken openParenToken, RegexToken openToken)
+        {
+            if ((openToken.Kind == RegexKind.LessThanToken && _currentToken.Kind == RegexKind.GreaterThanToken) ||
+                (openToken.Kind == RegexKind.SingleQuoteToken && _currentToken.Kind == RegexKind.SingleQuoteToken))
+            {
+                return _currentToken;
+            }
+
+            if (_currentToken.Kind == RegexKind.EndOfFile)
+            {
+                openParenToken = openParenToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Unrecognized_grouping_construct,
+                    GetSpan(openParenToken, openToken)));
+            }
+            else
+            {
+                openParenToken = openParenToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Invalid_group_name_Group_names_must_begin_with_a_word_character,
```

Comment:
The message seems odd. 
I would have expected an error saying something like "A closing token was expected", mentioning the proper closing token given the open token.

Reply:
So a core idea of the Roslyn regexp parser is that we emit the same diagnostic messages as hte .net one does.  This is the message they emit for this case, so we follow that.

This is an important invariant of the system (which we do lots of tests around) so that someone will not be confused by getting different error messages when they try execute their regex at runtime, versus when they edit their regex at design time.  I can add appropriate documentation about this in the class.

Reply:
Could you document this invariant?
Is that something that the tests check/enforce?

Reply:
> Could you document this invariant?

Yes.  this invariant is documented at top of the file.  However, i've expanded it with the following text:

```c#
    /// Note1: "report the same error" means that we will attempt to report the error using the same
    /// text the .net regex parser uses for its error messages.  This is so that the user is not
    /// confused when they use the IDE vs running the regex by getting different messages for the
    /// same issue.
```

> Is that something that the tests check/enforce?

Yes.  You can see that here: https://github.com/dotnet/roslyn/pull/25520/files#diff-4e08eaf0898f4a106091017995ed8cc2R167

We assert that the message the regex system emits contains the message we emit.

Reply:
Awesome. Thanks!

---
In reply to: [210051725](https://github.com/dotnet/roslyn/pull/23984#discussion_r210051725) [](ancestors = 210051725)